include/glow/Runtime/HostManager/HostManager.h

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_RUNTIME_HOSTMANAGER_H
+#define GLOW_RUNTIME_HOSTMANAGER_H
+#include "glow/Graph/Graph.h"
+#include "glow/Runtime/Executor.h"
+#include "glow/Runtime/Provisioner.h"
+#include "glow/Runtime/RuntimeTypes.h"
+
+#include <llvm/ADT/StringRef.h>
+
+#include <atomic>
+#include <map>
+#include <mutex>
+#include <unordered_map>
+#include <vector>
+
+namespace glow {
+namespace runtime {
+
+/// Callback signalling the result of running a function.
+using ResultCBTy = std::function<void(
+    runtime::RunIdentifierTy, runtime::ResultCode, std::unique_ptr<Context>)>;
+
+class DeviceManager;
+class Context;
+class DeviceConfig;
+class HostManager final : std::enable_shared_from_this<HostManager> {
+  /// Count of current in-flight networks being run. Atomic to allow concurrency
+  /// in runNetwork.
+  std::atomic<size_t> activeRequestCount_{0};
+
+  /// Count of total requests, this is used as a run ID. Atomic to allow
+  /// concurrency in runNetwork.
+  std::atomic<size_t> totalRequestCount_{0};
+
+  /// Limit maximum count of networks run at once. Hardcoded for now this should
+  /// be a configurable value.
+  unsigned int activeRequestLimit_ = 20;
+  /// A map from a networkName to the DAG that represents the network.
+  std::unordered_map<std::string, DAGNode> networks_;
+
+  /// Mutex for networks_ since runNetwork, addNetwork, and removeNetwork can
+  /// all be called concurrently a guard is needed.
+  std::mutex networkLock_;
+
+  /// A map of DeviceManagers by deviceID.
+  std::map<DeviceIDTy, std::unique_ptr<DeviceManager>> devices_;
+
+  /// Executor class, this handles dispatching execution requests to the
+  /// appropriate device managers for an inference request.
+  Executor executor_;
+
+  /// The provisioner assigns networks to specifc devices and handles
+  /// compilation for that device.
+  Provisioner provisioner_;
+
+public:
+  /// Adds the network to the host and does the necessary setup work. This
+  /// includes partitioning, provisioning, compiling and initializing backends.
+  /// Returns a result code to indicate success.
+  ResultCode addNetwork(Module *M);
```

Comment:
What are the pros to not passing ownership of Module here?

Reply:
Notionally the entity calling HostManager owns the module and persists, we can change this to pass ownership in and back out if desired.